### DISCUSSION TITLE
sisu-build-api

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-m2e/m2e-core#876] [General] [answered]
m2e uses [sisu-build-api](https://github.com/sonatype/sisu-build-api) to implement incremental builds, see https://www.eclipse.org/m2e/documentation/m2e-making-maven-plugins-compat.html

But looking at the repository, it is archive,d lat code changes where 12 years ago an it version still implies its some kind of experimental. Recently I found it might be good to enhance this API (e.g. for cases like https://github.com/eclipse-tycho/tycho/issues/1278) but this would of course only work with active development.

So I really wonder if:

1. m2e should adopt this project so we can improve on this API (might need to publish it under a new GAV and have some wrapper for the old API).
2. we think this could be useful for maven in general and try to let maven adopt this API, but then we give good advice to the maven people why we thing this is general purpose enough.

**Comment by mickaelistria** [comment]
This is a general Maven issue IMO; m2e is just 1 possible case for integration of build increments and Maven builds. So the solution shouldn't be part of m2e ideally. Although the BuildContext is somehow deprecated, I wonder whether Maven community does provide a good alternative already. It's definitely worth discussing it with Maven experts on the dev mailing-list.

**Reply by laeubi** [reply]
I have checked the maven "core" and it does not use `BuildContext`or similar, maybe [mvnd](https://github.com/apache/maven-mvnd) is a consumer but not sure about that, so m2e currently seems the only consumer of the API ... Also if I take a look at the git history it seems really closely related to m2e at the time there was development...

**Reply by HannesWell** [reply]
I agree that this is something that would be worth to discuss with the Maven devs.
IIRC for example the connector for the `maven-dependency-plugin`'s `copy` goal would not be necessary if that Plugin would use the mentiond build-context (or a successor). Probably this applies for other maven-plugins top.

**Reply by laeubi** [reply]
@HannesWell do you like to bring this to the mailing list and/or Jira?

**Reply by HannesWell** [reply]
I'm interested to do that, yes.
But I probably don't have time for that immediatly after my vacation next week.
If one of you wants to do it earlier I'm fine with IT and will follow interestedly and would participate I
if it fits.

**Reply by HannesWell** [reply]
> @HannesWell do you like to bring this to the mailing list and/or Jira?

Here we go:
https://www.mail-archive.com/dev@maven.apache.org/msg127887.html

**Reply by HannesWell** [reply]
> > @HannesWell do you like to bring this to the mailing list and/or Jira?
> 
> Here we go: https://www.mail-archive.com/dev@maven.apache.org/msg127887.html

Looks like their enthusiasm about adopting a 'BuildContext' like API by Maven is limited... Too bad.

**Comment by laeubi** [comment]
FYI @gnodet maybe also interesting for mvnd ...
@HannesWell any updates here?

**Reply by gnodet** [reply]
`mvnd` does not use any incremental build api.  What I've been focusing over the past years is rather than defining some kind of "incremental build api" that everyone would have to obey, have each plugin be able to optimise its processing by:
 * avoid any file overwrite if the content has not changed (using a [CachingOutputStream](https://github.com/codehaus-plexus/plexus-utils/blob/master/src/main/java/org/codehaus/plexus/util/io/CachingOutputStream.java) which does not touch files unless there's an actual change)
 * detect changes in the input (or rather, detect when there is no change)

With the two points implemented by most plugins, I've been able to have builds that do absolutely nothing (i.e. jars are not touched at all, etc...), so they are indeed much faster.   I _think_ this is roughly what the `BuildContext` intend to do, but it's also implemented for plain maven.  In short, if there's anything to help the plugin with, I think this should warrant a discussion, but I would think the API should belong in maven and that there should be a default working implementation in maven.  If there's a need to extend it for specific integration in IDEs, those IDEs could provide a different implementation.  But right now, there's an API which is sometimes used by plugins, that maven has no real knowledge of and no benefit.  Which I think indicates that it's bound to be removed from maven at some point in the future.

**Reply by laeubi** [reply]
The "detect changes in the input (or rather, detect when there is no change)" is especially something `BuildContext` provides (and abstracts) for plugins e.g. [BuildContext#hasDelta](https://github.com/sonatype/sisu-build-api/blob/master/src/main/java/org/sonatype/plexus/build/incremental/BuildContext.java#L45), and it gives some means to create streams (so it could be overplayed/optimized) as well ass adding markers to files (that is mostly for IDE).

The [default implementation](https://github.com/sonatype/sisu-build-api/blob/master/src/main/java/org/sonatype/plexus/build/incremental/DefaultBuildContext.java) is rather dump (and as maven-cli-build runs only once might not can be improved much), but e.g. mvnd (or m2e) can actually listen for a file-change, and then if the build is run one can ask the `BuildContext` if actually a file of interest has changed (without the need for the plugin to:

1. read the file
2. compare it so some previous state (where should this be stored?)

Still of course a plugin could decide that even if the file changed, actually nothing changed (e.g. just irrelevant white space added), so "incremental build" is a bit misleading, it is more like a server-push that tells the client (=Mojo) that there is a change in a file previously processed (or not).

I hope this makes it more clear and even if maven-core can not directly benefit from it I think many `maven-*-plugin`s can as they better fit in environments where a component can determine a diff e.g.:

1. IDEs usually know when a resource changes  compared to previous build (e.g. saving a source file)
2. `mvnd` or similar tools might use `WatchService` to get notified and then feed back that info to the `BuildContext`
3. CI-pipelines even can use the git-history to compute a delta between current/last (sucessfull) build and feed this info into a traditional maven run that only builds required module and feed back that info to the `BuildContext` so only mojos run where relevant data changed (that's actually planned for Tycho already).

But I think that the adoption of those can greatly benefit from having an official API (it must not be the sisu-build-api!) so it is just "standard" to implement and recommend for mojos to be used.

**Reply by gnodet** [reply]
> The "detect changes in the input (or rather, detect when there is no change)" is especially something `BuildContext` provides (and abstracts) for plugins e.g. [BuildContext#hasDelta](https://github.com/sonatype/sisu-build-api/blob/master/src/main/java/org/sonatype/plexus/build/incremental/BuildContext.java#L45), and it gives some means to create streams (so it could be overplayed/optimized) as well ass adding markers to files (that is mostly for IDE).
> 
> The [default implementation](https://github.com/sonatype/sisu-build-api/blob/master/src/main/java/org/sonatype/plexus/build/incremental/DefaultBuildContext.java) is rather dump (and as maven-cli-build runs only once might not can be improved much), but e.g. mvnd (or m2e) can actually listen for a file-change, and then if the build is run one can ask the `BuildContext` if actually a file of interest has changed (without the need for the plugin to:
> 
> 1. read the file
> 2. compare it so some previous state (where should this be stored?)

Lots of plugins do store the state in order to skip unneeded execution.  So that very part should be provided by the default implementation.  The information is usually stored in the `target` folder, which means a `clean` will remove the information, but I don't see that as a big problem.

> 
> Still of course a plugin could decide that even if the file changed, actually nothing changed (e.g. just irrelevant white space added), so "incremental build" is a bit misleading, it is more like a server-push that tells the client (=Mojo) that there is a change in a file previously processed (or not).
> 
> I hope this makes it more clear and even if maven-core can not directly benefit from it I think many `maven-*-plugin`s can as they better fit in environments where a component can determine a diff e.g.:
> 
> 1. IDEs usually know when a resource changes  compared to previous build (e.g. saving a source file)
> 2. `mvnd` or similar tools might use `WatchService` to get notified and then feed back that info to the `BuildContext`
> 3. CI-pipelines even can use the git-history to compute a delta between current/last (sucessfull) build and feed this info into a traditional maven run that only builds required module and feed back that info to the `BuildContext` so only mojos run where relevant data changed (that's actually planned for Tycho already).
> 
> But I think that the adoption of those can greatly benefit from having an official API (it must not be the sisu-build-api!) so it is just "standard" to implement and recommend for mojos to be used.

I disagree.  I think that's a very bad idea to have plugins written with 2 different code paths whether there is a `BuildContext` that will provide some information or not.  A usable and useful implementation of the `BuildContext` would allow migrating the plugins to a single codepath that could be maintainable.

**Reply by laeubi** [reply]
> Lots of plugins do store the state in order to skip unneeded execution. So that very part should be provided by the default implementation. The information is usually stored in the `target` folder, which means a `clean` will remove the information, but I don't see that as a big problem.

Yes something like that would be possible, just that not every plugin has implement storage / retrieval itself

> I disagree. I think that's a very bad idea to have plugins written with 2 different code paths whether there is a `BuildContext` that will provide some information or not. A usable and useful implementation of the `BuildContext` would allow migrating the plugins to a single codepath that could be maintainable.

Possibly we have a misunderstanding here, there won't be two code paths, a `BuildContext` will (and is) always there already, the current (dumb) default simply claims that a file always has changed what makes the default a no-op (no delta compared to previous build), but when run in m2e the default impl is replaced with one that knows about the workspace state an what files has changed in this run.

So for the plugin there is no difference at all and it should/would never know or care about.

**Reply by gnodet** [reply]
> > Lots of plugins do store the state in order to skip unneeded execution. So that very part should be provided by the default implementation. The information is usually stored in the `target` folder, which means a `clean` will remove the information, but I don't see that as a big problem.
> 
> Yes something like that would be possible, just that not every plugin has implement storage / retrieval itself
> 
> > I disagree. I think that's a very bad idea to have plugins written with 2 different code paths whether there is a `BuildContext` that will provide some information or not. A usable and useful implementation of the `BuildContext` would allow migrating the plugins to a single codepath that could be maintainable.
> 
> Possibly we have a misunderstanding here, there won't be two code paths, a `BuildContext` will (and is) always there already, the current (dumb) default simply claims that a file always has changed what makes the default a no-op (no delta compared to previous build), but when run in m2e the default impl is replaced with one that knows about the workspace state an what files has changed in this run.
> 
> So for the plugin there is no difference at all and it should/would never know or care about.

That would be true only for plugins that accept to not optimise their processing.  If the plugin wants to do some stale files detection or any similar optimisation, it will have to work on two different code paths.

In short, I think that there is a need for such an API, but it has to be defined in the maven project, has to actually provide some real value for non IDE builds, and that can be integrated by other IDEs as well (not only Eclipse).    There are been other attempts like [takari incremental build](https://github.com/takari/io.takari.incrementalbuild/).  Anyway, this discussion should be brought to `dev@maven.apache.org` imho.

**Reply by laeubi** [reply]
> That would be true only for plugins that accept to not optimise their processing. If the plugin wants to do some stale files detection or any similar optimisation, it will have to work on two different code paths.

The only difference here is, that one adds an additional check for example:

```
if (isMyFileStale(input, output) {
  //do something
}
```

will become

```
if (buildContext.hasDelta(input) && isMyFileStale(input, output) {
  //do something
}
```

so no need for different code path...

> In short, I think that there is a need for such an API, but it has to be defined in the maven project, has to actually provide some real value for non IDE builds, 

That's why I though `mvnd` would be another consumer

> and that can be integrated by other IDEs as well (not only Eclipse). 

The current API can already be used by anyone (not only IDEs) and there are already some plugins out there using that.

>There are been other attempts like [takari incremental build](https://github.com/takari/io.takari.incrementalbuild/). 

The takiri one seem to adress a whole new framework but I might be wrong.

> Anyway, this discussion should be brought to `dev@maven.apache.org` imho.

I hope this will be done soon, but its always good to first collect some feedback from a smaller audience.

**Reply by gnodet** [reply]
> > That would be true only for plugins that accept to not optimise their processing. If the plugin wants to do some stale files detection or any similar optimisation, it will have to work on two different code paths.
> 
> The only difference here is, that one adds an additional check for example:
> 
> ```
> if (isMyFileStale(input, output) {
>   //do something
> }
> ```
> 
> will become
> 
> ```
> if (buildContext.hasDelta(input) && isMyFileStale(input, output) {
>   //do something
> }
> ```
> 
> so no need for different code path...

I think stale detection should be provided by the build context (or whatever the api is).  Maybe I'm wrong, but the whole point of this API is to detect which files need to be processed, so this looks clearly redundant to me.

> 
> > In short, I think that there is a need for such an API, but it has to be defined in the maven project, has to actually provide some real value for non IDE builds,
> 
> That's why I though `mvnd` would be another consumer

Yes, it may be.  But if the state is stored, I'm not sure there's much benefit for `mvnd` to implement its own.  Actually, if done properly, I'm not sure what Eclipse would actually gain.  There's not much difference in flagging through a file watcher or checking the files directly and comparing with the output.

> 
> > and that can be integrated by other IDEs as well (not only Eclipse).
> 
> The current API can already be used by anyone (not only IDEs) and there are already some plugins out there using that.

Sure, but it would be nice to make sure the API covers all the needs.

> 
> > There are been other attempts like [takari incremental build](https://github.com/takari/io.takari.incrementalbuild/).
> 
> The takiri one seem to adress a whole new framework but I might be wrong.

Even for plugins, the API may miss a few things.  The api needs to be able to store some configuration bits that will affect the processing, the api for deleted resources does not seem straightforward, and the one-to-one mapping between input and output might be too limitative.

> > Anyway, this discussion should be brought to `dev@maven.apache.org` imho.
> 
> I hope this will be done soon, but its always good to first collect some feedback from a smaller audience.

**Reply by HannesWell** [reply]
> @HannesWell any updates here?

Not yet unfortunately. If you want it to happen faster I wouldn't mind if you take care of it otherwise I can do that hopefully not too far in the future.

Besides the aspect of avoiding unnecessary file modifications the `BuildContext` also has the purpose to notify Eclipse (or the IDE in general, but I'm not aware how other IDEs handle that) about files/folders that have been modified and should therefore be refreshed. Otherwise the changes do not become visible to the Resources-Plugin of Eclipse and Eclipse can not show them to the user or consider them if necessary.

For example the only thing the connector for the `maven-dependency-plugin` does is to obtain the target directories of the copy/unpack goal and to tell Eclipse to refresh them:
https://github.com/ianbrandt/m2e-maven-dependency-plugin/blob/38c09f1ee625369f5ebc7a57f5b2a63e9034f9a4/core/src/com/ianbrandt/tools/m2e/mdp/core/MdpBuildParticipant.java#L71

If the `maven-dependency-plugin` would use `BuildContext.newFileOutputStream()` or something similar in a more general API then that whole connector would not be necessary.
The same applies for the `maven-clean-plugin`, for which no connector exists yet (as far as I know). But if there would be a method like `BuildContext.delete()` no connector would be necessary as well to make the IDE aware of additional directories that are cleaned by the clean goal.

**Reply by laeubi** [reply]
@HannesWell maybe we should propose the usage of the current BuildContext to these plugins with a lifecycle mapping xml? That should work as of today already...

**Reply by HannesWell** [reply]
> @HannesWell maybe we should propose the usage of the current BuildContext to these plugins with a lifecycle mapping xml? That should work as of today already...

Yes that should work already. But my guess is that if we propose to use such abandoned prototype API the maintainers of those plugins will ask for something more robust which will lead to the general discussion, so I think we can just start that.
Nevertheless those two plug-ins are good examples.
It is probably faster if we simply add connectors for them to m2e. Maybe in a new plugin `o.e.m2e.standard.connector` or similar.

I started to prepare a little text for the Maven dev-mailing list that I hopefully complete in the next days and that I then plan to post on the Maven list. I will add a link here, but I have seen you are already on that list so will notice that any ways. :)

**Comment by laeubi** [comment, marked as answer]
The API is now revived in a new repository:

https://github.com/codehaus-plexus/plexus-build-api

given the reserved feedback from the maven-dev it seems most suitable to help with improving the API keeping it alive and healthy and hope for some glory future of maven 4 (or 5 ...), I opened some issues so everyone is invited to help:

https://github.com/codehaus-plexus/plexus-build-api/issues

**Reply by HannesWell** [reply]
Since the new plexus-build-api has new a new package name we can support both for some time to allow transition.

Too bad the new API was released already without your suggestions being considered. But from my quick overview I suspect that some of them are breaking changes so they cannot be applied for version 1.x.

This also makes me aware that it will be hard to support multiple major/incompatible versions of that API (if it evolves one day). Theoretically it would work if we have one impl bundle/project for each major version and detect which version a plugin uses, but that will be somewhat cumbersome, but theoretically doable.

**Reply by laeubi** [reply]
I'm not aware of any release, but we most likely will only support it up from version X where X is a version we are fine with and then only latest version that will hopefully be backward compatible for clients.

**Reply by HannesWell** [reply]
There was a release eight days ago:
https://github.com/codehaus-plexus/plexus-build-api/releases/tag/plexus-build-api-1.0.0



> I'm not aware of any release, but we most likely will only support it up from version X where X is a version we are fine with and then only latest version that will hopefully be backward compatible for clients.

Yep, that's what I also had in mind.

**Reply by HannesWell** [reply]
For reference, supporting the new BuildContext-API is tracked in #944.
